tensorflow/core/kernels/whole_file_read_ops.cc

```diff
@@ -119,4 +119,34 @@ class ReadFileOp : public OpKernel {
 
 REGISTER_KERNEL_BUILDER(Name("ReadFile").Device(DEVICE_CPU), ReadFileOp);
 
+class WriteFileOp : public OpKernel {
+  public:
+    using OpKernel::OpKernel;
+    void Compute(OpKernelContext* context) override {
+	    const Tensor* filename_input;
+	    const Tensor* contents_input;
+	    Tensor* output = nullptr;
+	    OP_REQUIRES_OK(context, context->input("filename", &filename_input));
+	    OP_REQUIRES_OK(context, context->input("contents", &contents_input));
+	    OP_REQUIRES(context, TensorShapeUtils::IsScalar(filename_input->shape()),
+			    errors::InvalidArgument(
+				    "Input filename tensor must be scalar, but had shape: ",
+				    filename_input->shape().DebugString()));
+	    OP_REQUIRES(context, TensorShapeUtils::IsScalar(contents_input->shape()),
+			    errors::InvalidArgument(
+				    "Contents tensor must be scalar, but had shape: ",
+				    contents_input->shape().DebugString()));
+	    OP_REQUIRES_OK(context, context->allocate_output("output", TensorShape({}),
+				    &output));
+	    if(WriteStringToFile(context->env(),
+				    filename_input->scalar<string>()(),
+				    contents_input->scalar<string>()()).ok()) {
+	    	output->scalar<int>()(0) = 0;
+	    } else {
+	    	output->scalar<int>()(0) = -1;
```

Comment:
Take a look at the shape functions of `MergeV2Checkpoints` and `Save`: both have no outputs, so their shape fns just enforce some properties about the inputs.

The one in the Python side does not change -- it just means "use the shape fn registered in C++'s side".